tools/schemacode/bidsschematools/validator.py

```diff
@@ -1,31 +1,139 @@
 import datetime
+import fnmatch
 import json
 import os
 import re
-import typing as ty
-from collections.abc import Mapping
 from copy import deepcopy
-from functools import lru_cache
 from pathlib import Path
 
 import bidsschematools as bst
+import bidsschematools.parse
 import bidsschematools.schema
 import bidsschematools.types
 import bidsschematools.utils
 
 lgr = bst.utils.get_logger()
 
-# The list of which entities create directories could be dynamically specified by the YAML, but for
-# now, it is not.
-# Ordering is important, as "subject" follows "session" alphabetically, but is hierarchically
-# above it.
-DIR_ENTITIES = ["subject", "session"]
+VALIDATOR_SCHEMA_COMPATIBILITY_LEVEL = "minor"
+
+
+def _bids_schema_versioncheck(schema_dir, compatibility=VALIDATOR_SCHEMA_COMPATIBILITY_LEVEL):
+    """
+    Use the `SCHEMA_VERSION` descriptor file to determine whether a selected schema directory is
+    compatible with the validator.
+
+    Parameters
+    ----------
+    schema_dir : str
+        A string which specifies a path.
+    compatibility : "major" or "minor", optional
+        Compatibility range of the validator. Major means "M.*" are considered compatible, minor
+        means "M.m.*" are considered compatible.
+
+    Returns
+    -------
+    bool:
+        Whether the schema is compatible with the validator.
+    """
+
+    if compatibility not in ("major", "minor"):
+        raise ValueError("Schema compatibility needs to be set to either “major” or “minor”.")
+
+    schema_version_file = os.path.join(schema_dir, "SCHEMA_VERSION")
+    try:
+        with open(schema_version_file, "r") as f:
+            schema_version = f.readlines()[0].strip()
+    except FileNotFoundError:
+        lgr.warning(
+            "The selected schema directory, `%s`, does not contain a SCHEMA_VERSION file. "
+            "Cannot ascertain compatibility. Attempting to query the BIDS reference for "
+            "compatible versions.",
+            schema_dir,
+        )
+    else:
+        nparts = 1 if compatibility == "major" else 2
+        return schema_version.split(".", nparts)[:-1] == bst.__version__.split(".", nparts)[:-1]
```

Comment:
We should be able to test this with something like:

```Python
def test_incompatible_version(monkeypatch):
    schema_dir = get_bundled_schema_dir()
    assert _bids_schema_versioncheck(schema_dir)
    monkeypatch.setattr(bst, '__version__', '99.99.99')
    assert not _bids_schema_versioncheck(schema_dir)

Reply:
done in [0d9ee22](https://github.com/bids-standard/bids-specification/pull/1398/commits/0d9ee22a5afbb6627e56dcdcc8c7dfca700e927a)